Bluetooth: controller: Fix incorrect master role scheduling

cvinayak opened the pull request:
Revert incorrect calculation introduced in
commit ec5a787da2ab ("Bluetooth: controller: Fix multiple
master role event scheduling") and revert a related
incorrect fix in commit a02606cbf9d2 ("Bluetooth:
controller: Fix missing ticks to us conversion").

Relates to: #5486

Fixes the controller assert in ctrl.c line number 1477. A
64-bit arithmetic took ~35 us in Radio ISR for nRF51 causing
the ISR to take too much time before packet buffer could be
set.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix incorrect master role scheduling

Revert incorrect calculation introduced in
commit ec5a787da2ab ("Bluetooth: controller: Fix multiple
master role event scheduling") and revert a related
incorrect fix in commit a02606cbf9d2 ("Bluetooth:
controller: Fix missing ticks to us conversion").

Fixes the controller assert in ctrl.c line number 1477. A
64-bit arithmetic took ~35 us in Radio ISR for nRF51 causing
the ISR to take too much time before packet buffer could be
set.

Also, fixed master scheduling by correctly accounting for
the jitter between each master event.

Relates to: #5486

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

jhedberg commented:
@cvinayak is this now considered a serious fix candidate? (I'd assume so since you removed "RFC" from the subject). Are you waiting for some more testing to happen?

carlescufi commented:
@cvinayak can you rebase and fix conflicts?

cvinayak commented:
@carlescufi and @jhedberg Do not merge, I may be onto something with a connection failed to be established issue when using window offsets. I need to analyze further.

cvinayak committed:
Bluetooth: controller: Use XON_BITMASK instead of BIT(31)

Use XON_BITMASK define instead of raw BIT(31) or bit shift
operations in code to represent use of reduced prepared by
an event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Define HAL_TIMER_SIGN_BIT

Define a macro HAL_TIMER_SIGN_BIT to correspond to the most
significant bit support by counter hardware used by ticker.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix XTAL_ADVANCED threshold detection

Fix incorrect calculation of unreserved timespace which
should take into account the ticks_slot of current event,
and compensation for any reduced prepare in the current
event as well as in the next event.

This regression relates to the commit ad7c9d3d76ae
("Bluetooth: controller: Improved continuous scanning")

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@jhedberg and @MariuszSkamra could you please recheck with PTS (relates to nRF52 failed to establish connection as central, and nRF51 asserted as central).

jhedberg commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -4776,115 +4798,97 @@ static void mayfly_xtal_stop_calc(void *params)
+	ticks_prepare_to_start_next = (hdr_next->ticks_active_to_start >
+				       hdr_next->ticks_preempt_to_start) ?
+				      hdr_next->ticks_active_to_start :
+				      hdr_next->ticks_preempt_to_start;
```
`ticks_prepare_to_start_next = max(hdr_next->ticks_active_to_start, hdr_next->ticks_preempt_to_start);`

jhedberg commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -4776,115 +4798,97 @@ static void mayfly_xtal_stop_calc(void *params)
+			(hdr_curr->ticks_active_to_start >
+			 hdr_curr->ticks_preempt_to_start) ?
+			hdr_curr->ticks_active_to_start :
+			hdr_curr->ticks_preempt_to_start;
```
`ticks_prepare_to_start = max(hdr_next->ticks_active_to_start, hdr_next->ticks_preempt_to_start);`

jhedberg reviewed (commented):
A couple of minor style clarifications since we do have min & max macros that should be preferred whenever possible. I realize the issue was already in the old code, so it's fine to do the cleanup in later patches as well.

jhedberg commented:
Both nRF51 and nRF52 seem to work ok with the latest patches in this PR

MariuszSkamra commented:
I confirm this fixes #5486 and #4989 running combined build on nrf52840 board.

1/4   GATT   GATT/CL/GAW/BV-02-C   PASS
2/4   GATT   GATT/SR/GAW/BI-01-C   PASS
3/4   GATT   GATT/SR/GAS/BV-01-C   PASS
4/4   GATT   GATT/SR/GAW/BV-02-C   PASS

MariuszSkamra commented:
Tested - works!

carlescufi commented:
@cvinayak should we merge this?

cvinayak commented:
@MariuszSkamra Thank you.
@carlescufi I will run some conformance tests tonight and some manual testing tomorrow morning, want to self review some more, last time I touched this scheduling code I messed it up bad.

cvinayak commented:
@MariuszSkamra, @jhedberg and @carlescufi 

All selected 283 of 283 LL.TS.5.0.2 conformance tests passed!

Feel free to merge.
